### PR TITLE
Decimal support for Number

### LINKED ISSUE #48: Add Decimal to Clay.Number
`Clay.Number` is a union type, useful for scenarios (such as `System.Json`) where numeric fidelity is important but the system may return 1 of several different types of numerics. 

In specific use cases we need to support `Decimal` as well as the `int` and `real` types already supported.
This will bloat the size of the union unfortunately.

### PR DESCRIPTION
**Pull request by jcdickinson**
Adds decimal support to number, as well as some cleanups.

- Adding decimal type to Number.
- Adding ToDecimal method.
- Returning field directly in conversion methods, when the type matches.
- Fixing some bugs in the Number tests.

Fixes #48.

### CODE REVIEW

File: src/SourceCode.Clay.Tests/NumberTests.cs [hunk 1]
```diff
@@ -30,11 +30,15 @@ public class NumberTests
...
             InlineData(nameof(Single), 9.0F, NumberKind.Real | NumberKind.Signed, false),
             InlineData(nameof(Double) + "=0", 0.0, NumberKind.Real | NumberKind.Signed, true),
-            InlineData(nameof(Double), 10.0, NumberKind.Real | NumberKind.Signed, false)
+            InlineData(nameof(Double), 10.0, NumberKind.Real | NumberKind.Signed, false),
+            InlineData(nameof(Decimal), 10.0, NumberKind.Real | NumberKind.Signed, false)
         ]
         [Theory(DisplayName = nameof(Number_ContructGet))]
         public void Number_ContructGet(string description, object expected, NumberKind kind, bool isZero)
         {
+            if (description == nameof(Decimal))
```
Comment (jcdickinson) [inline src/SourceCode.Clay.Tests/NumberTests.cs]: Decimal literals can't be used in attributes, so we need to convert in the test.

File: src/SourceCode.Clay/Number.cs [hunk 2]
```diff
@@ -767,18 +852,27 @@ public int CompareTo(Number other)
...
                 case (((uint)TypeCode.Single) << 5) | (uint)TypeCode.Int32: return _single.CompareTo(other._int32);
                 case (((uint)TypeCode.Single) << 5) | (uint)TypeCode.UInt32: return _single < 0 ? -1 : _single.CompareTo(other._uint32);
-                case (((uint)TypeCode.Single) << 5) | (uint)TypeCode.Int64: return ((double)_single).CompareTo((double)other._int64);
-                case (((uint)TypeCode.Single) << 5) | (uint)TypeCode.UInt64: return _single < 0 ? -1 : ((double)_single).CompareTo((double)other._uint64);
+                case (((uint)TypeCode.Single) << 5) | (uint)TypeCode.Int64: return ((double)_single).CompareTo(other._int64);
+                case (((uint)TypeCode.Single) << 5) | (uint)TypeCode.UInt64: return _single < 0 ? -1 : ((double)_single).CompareTo(other._uint64);
                 case (((uint)TypeCode.Single) << 5) | (uint)TypeCode.Single: return _single.CompareTo(other._single);
                 case (((uint)TypeCode.Single) << 5) | (uint)TypeCode.Double: return -other._double.CompareTo(_single);
+                case (((uint)TypeCode.Single) << 5) | (uint)TypeCode.Decimal:
```
Comment (jcdickinson) [inline src/SourceCode.Clay/Number.cs]: Single and Double can contain values that Decimal cannot.

File: src/SourceCode.Clay.Tests/NumberTests.cs [hunk 3]
```diff
@@ -30,11 +30,15 @@ public class NumberTests
             InlineData(nameof(Single) + "=0", 0.0F, NumberKind.Real | NumberKind.Signed, true),
             InlineData(nameof(Single), 9.0F, NumberKind.Real | NumberKind.Signed, false),
             InlineData(nameof(Double) + "=0", 0.0, NumberKind.Real | NumberKind.Signed, true),
-            InlineData(nameof(Double), 10.0, NumberKind.Real | NumberKind.Signed, false)
+            InlineData(nameof(Double), 10.0, NumberKind.Real | NumberKind.Signed, false),
+            InlineData(nameof(Decimal), 10.0, NumberKind.Real | NumberKind.Signed, false)
         ]
         [Theory(DisplayName = nameof(Number_ContructGet))]
         public void Number_ContructGet(string description, object expected, NumberKind kind, bool isZero)
```
Comment (grant-d) [inline src/SourceCode.Clay.Tests/NumberTests.cs]: `static` if possible

File: src/SourceCode.Clay.Tests/NumberTests.cs (hunk 1 above)
Comment (grant-d) [reply src/SourceCode.Clay.Tests/NumberTests.cs]: Else maybe create actual `const` decimals in class

File: src/SourceCode.Clay/Number.cs [hunk 4]
```diff
@@ -1034,37 +1190,38 @@ int IConvertible.ToInt32(IFormatProvider provider)
                 case TypeCode.Byte: return ((IConvertible)_byte).ToInt32(provider);
```
Comment (grant-d) [inline src/SourceCode.Clay/Number.cs]: Can we not do direct conversion on this?

**Pull request merged**